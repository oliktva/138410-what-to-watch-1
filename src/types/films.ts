import PropTypes from 'prop-types';

export const filmPropTypes = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  img: PropTypes.string.isRequired,
  video: PropTypes.string.isRequired
});

export const filmsPropTypes = PropTypes.arrayOf(filmPropTypes.isRequired);

export interface FilmProps {
  name: string;
  img: string;
  video: string;
  id: number;
}
