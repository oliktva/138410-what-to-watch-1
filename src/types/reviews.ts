export interface ReviewProps {
  comment: string;
  date: string;
  id: number;
  rating: number;
  user: {
    id: number;
    name: string;
  };
}
