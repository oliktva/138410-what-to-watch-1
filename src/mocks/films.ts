import {FilmProps} from "src/types/films";

export const films: FilmProps[] = [{
  id: 1,
  name: `Fantastic Beasts`,
  img: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 2,
  name: `Bohemian Rhapsody`,
  img: `img/bohemian-rhapsody.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 3,
  name: `Macbeth`,
  img: `img/macbeth.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 4,
  name: `Aviator`,
  img: `img/aviator.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 5,
  name: `We need to talk about Kevin`,
  img: `img/we-need-to-talk-about-kevin.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 6,
  name: `What We Do in the Shadows`,
  img: `img/what-we-do-in-the-shadows.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 7,
  name: `Revenant`,
  img: `img/revenant.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 8,
  name: `Johnny English`,
  img: `img/johnny-english.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 9,
  name: `Shutter Island`,
  img: `img/shutter-island.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 10,
  name: `Pulp Fiction`,
  img: `img/pulp-fiction.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 11,
  name: `No Country for Old Men`,
  img: `img/no-country-for-old-men.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 12,
  name: `Snatch`,
  img: `img/snatch.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 13,
  name: `Moonrise Kingdom`,
  img: `img/moonrise-kingdom.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 14,
  name: `Seven Years in Tibet`,
  img: `img/seven-years-in-tibet.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 15,
  name: `Midnight Special`,
  img: `img/midnight-special.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 16,
  name: `War of the Worlds`,
  img: `img/war-of-the-worlds.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 17,
  name: `Dardjeeling Limited`,
  img: `img/dardjeeling-limited.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 18,
  name: `Orlando`,
  img: `img/orlando.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 19,
  name: `Mindhunter`,
  img: `img/mindhunter.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 20,
  name: `Midnight Special`,
  img: `img/midnight-special.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}];
