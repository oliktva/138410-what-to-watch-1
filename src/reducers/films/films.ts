import {Action as ReduxAction} from 'redux';
import {ThunkDispatch as ReduxThunkDispatch, ThunkAction as ReduxThunkAction} from 'redux-thunk';
import {AxiosResponse, AxiosInstance} from 'axios';
import {toCamel} from 'convert-keys';

import {GenreProps} from 'src/types/genres';
import {FilmProps} from 'src/types/films';
import {ReviewProps} from 'src/types/reviews';
import {ThunkDispatch, ThunkAction, State as AppState} from 'src/types/reducer';

export const SET_GENRE = `SET_GENRE`;
export const LOAD_FILMS_SUCCESS = `LOAD_FILMS_SUCCESS`;
export const LOAD_FAVORITE_SUCCESS = `LOAD_FAVORITE_SUCCESS`;
export const UPDATE_FILM = `UPDATE_FILM`;
export const LOAD_PROMO = `LOAD_PROMO`;
export const LOAD_REVIEWS = `LOAD_REVIEWS`;

const ALL_GENRES = `All genres`;

export interface State {
  genre: GenreProps;
  items: FilmProps[];
  favorite: FilmProps[];
  promo: FilmProps | null;
  reviews: {[key: number]: ReviewProps[]} | null;
}

interface SetFilterAction extends ReduxAction {
  type: typeof SET_GENRE;
  payload: GenreProps;
}

interface LoadFilmsAction extends ReduxAction {
  type: typeof LOAD_FILMS_SUCCESS;
  payload: FilmProps[];
}

interface LoadFavoritesAction extends ReduxAction {
  type: typeof LOAD_FAVORITE_SUCCESS;
  payload: FilmProps[];
}

interface UpdateFilmAction extends ReduxAction {
  type: typeof UPDATE_FILM;
  payload: FilmProps;
}

interface LoadFilmPromoAction extends ReduxAction {
  type: typeof LOAD_PROMO;
  payload: FilmProps;
}

interface LoadReviewsAction extends ReduxAction {
  type: typeof LOAD_REVIEWS;
  payload: {id: number; reviews: ReviewProps[]};
}

export type Action =
  SetFilterAction | LoadFilmsAction | LoadFavoritesAction | UpdateFilmAction | LoadFilmPromoAction | LoadReviewsAction;

export type ThunkDispatch = ReduxThunkDispatch<State, AxiosInstance, Action>;
export type ThunkAction = ReduxThunkAction<Promise<void>, State, AxiosInstance, Action>;

export const initialState: State = {
  genre: ALL_GENRES,
  items: [],
  favorite: [],
  promo: null,
  reviews: null
};

const updateFilm = (film: FilmProps, filmsList: FilmProps[]): FilmProps[] =>
  filmsList.map((f: FilmProps): FilmProps => (f.id === film.id ? film : f));

export const ActionCreator = {
  setFilterByGenre: (genre: GenreProps): SetFilterAction => {
    return {
      type: SET_GENRE,
      payload: genre
    };
  },

  loadFilms: (films: FilmProps[]): LoadFilmsAction => {
    return {
      type: LOAD_FILMS_SUCCESS,
      payload: films
    };
  },

  loadFavorite: (films: FilmProps[]): LoadFavoritesAction => {
    return {
      type: LOAD_FAVORITE_SUCCESS,
      payload: films
    };
  },

  updateFilm: (film: FilmProps): UpdateFilmAction => {
    return {
      type: UPDATE_FILM,
      payload: film
    };
  },

  loadPromo: (film: FilmProps): LoadFilmPromoAction => {
    return {
      type: LOAD_PROMO,
      payload: film
    };
  },

  loadReviews: (id: number, reviews: ReviewProps[]): LoadReviewsAction => {
    return {
      type: LOAD_REVIEWS,
      payload: {id, reviews}
    };
  }
};

export const Operation = {
  loadFilms: (): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.get(`/films`)
        .then((response: AxiosResponse<Record<string, any>[]>): void => {
          const data = response.data.map((f): FilmProps => toCamel<FilmProps>(f));

          dispatch(ActionCreator.loadFilms(data));
        });
    };
  },

  loadFavorite: (): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.get(`/favorite`)
        .then((response: AxiosResponse<Record<string, any>[]>): void => {
          const data = response.data.map((f): FilmProps => toCamel<FilmProps>(f));

          dispatch(ActionCreator.loadFavorite(data));
        });
    };
  },

  addToFavorites: (filmId: number): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.post(`/favorite/${filmId}/1`)
        .then((response: AxiosResponse<Record<string, any>>): void => {
          const data = toCamel<FilmProps>(response.data);

          dispatch(ActionCreator.updateFilm(data));
        });
    };
  },

  removeFromFavorites: (filmId: number): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.post(`/favorite/${filmId}/0`)
        .then((response: AxiosResponse<Record<string, any>>): void => {
          const data = toCamel<FilmProps>(response.data);

          dispatch(ActionCreator.updateFilm(data));
        });
    };
  },

  loadPromo: (): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.get(`/films/promo`)
        .then((response: AxiosResponse<Record<string, any>>): void => {
          const data = toCamel<FilmProps>(response.data);

          dispatch(ActionCreator.loadPromo(data));
        });
    };
  },

  loadReviews: (id: number): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.get(`/comments/${id}`)
        .then((response: AxiosResponse<Record<string, any>[]>): void => {
          const data = response.data.map((r): ReviewProps => toCamel<ReviewProps>(r));

          dispatch(ActionCreator.loadReviews(id, data));
        });
    };
  },

  addReview: (id: number, reviewText: string, rating: string | number): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.post(`/comments/${id}`, {
        comment: reviewText,
        rating: typeof rating === `string` ? parseInt(rating, 10) : rating
      }).then((response: AxiosResponse<Record<string, any>[]>): void => {
        const data = response.data.map((r): ReviewProps => toCamel<ReviewProps>(r));

        dispatch(ActionCreator.loadReviews(id, data));
      });
    };
  }
};

const reducer = (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case LOAD_FILMS_SUCCESS:
      return {
        ...state,
        items: action.payload
      };

    case SET_GENRE:
      return {
        ...state,
        genre: action.payload
      };

    case LOAD_FAVORITE_SUCCESS:
      return {
        ...state,
        favorite: action.payload
      };

    case UPDATE_FILM:
      return {
        ...state,
        promo: state.promo ? updateFilm(action.payload, [state.promo])[0] : state.promo,
        items: updateFilm(action.payload, state.items),
        favorite: updateFilm(action.payload, state.favorite)
      };

    case LOAD_PROMO: {
      return {
        ...state,
        promo: action.payload
      };
    }

    case LOAD_REVIEWS: {
      return {
        ...state,
        reviews: {
          ...state.reviews,
          [action.payload.id]: action.payload.reviews
        }
      };
    }

    default:
      return state;
  }
};

export default reducer;
