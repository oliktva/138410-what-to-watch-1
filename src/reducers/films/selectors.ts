import {createSelector} from 'reselect';

import Namespaces from 'src/reducers/namespaces';

import {State} from 'src/types/reducer';
import {FilmProps} from 'src/types/films';
import {GenreProps} from 'src/types/genres';

const ALL_GENRES = `All genres`;

export const getFilms = (state: State): FilmProps[] => {
  return state[Namespaces.FILMS].items;
};

export const getGenre = (state: State): GenreProps => {
  return state[Namespaces.FILMS].genre;
};

export const getFilmsByGenre = createSelector(
  getFilms,
  getGenre,
  (films: FilmProps[], genre: GenreProps): FilmProps[] => {
    return genre === ALL_GENRES ?
      films :
      films.filter((f: FilmProps): boolean => f.genre === genre);
  }
);

export const getGenres = createSelector(
  getFilms,
  (films: FilmProps[]): GenreProps[] => {
    const filmsGenres = films.map((f: FilmProps): GenreProps => f.genre)
      .reduce((result: GenreProps[], genre: GenreProps): GenreProps[] => {
        if (!result.includes(genre)) {
          result.push(genre);
        }

        return result;
      }, []);

    return [ALL_GENRES, ...filmsGenres];
  }
);
