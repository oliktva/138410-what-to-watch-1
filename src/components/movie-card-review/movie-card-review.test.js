import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import MovieCardReview from './movie-card-review';
import {film} from 'src/fixtures/films';
import {review} from 'src/fixtures/reviews';

Enzyme.configure({adapter: new Adapter()});

it(`renders correctly`, () => {
  const movieCardReview = shallow(<MovieCardReview
    header={<div />}
    film={film}
    userReview={review}
    addReview={jest.fn()}
  />);

  expect(toJson(movieCardReview)).toMatchSnapshot();
});
