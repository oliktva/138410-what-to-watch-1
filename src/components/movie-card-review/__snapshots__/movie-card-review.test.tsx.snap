// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Fragment>
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <BackgroundImage
        image="https://es31-server.appspot.com/wtw/static/film/background/Pulp_Fiction.jpg"
        name="Pulp Fiction"
      />
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <div />
      <MoviePoster
        image="https://es31-server.appspot.com/wtw/static/film/poster/Pulp_Fiction.jpg"
        kind="small"
        name="Pulp Fiction"
      />
    </div>
    <WithFormFields
      addReview={[MockFunction]}
      id={1}
      userReview={
        Object {
          "comment": "This is comment",
          "date": "54367894",
          "id": 1,
          "rating": 8,
          "user": Object {
            "id": 2,
            "name": "Joe",
          },
        }
      }
    />
  </section>
</Fragment>
`;
